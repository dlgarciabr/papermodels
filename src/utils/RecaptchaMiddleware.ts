import { MiddlewareNext, MiddlewareResponse, RequestMiddleware } from 'blitz';
import { IncomingMessage } from 'http';

interface IRequestMiddlewareRequest extends IncomingMessage {
  body: any;
  cookies: any;
  query: any;
}

interface IRequestMiddlewareResponse extends MiddlewareResponse {
  status: Function;
}

const protectedUrls = ['/api/rpc/getItemAnonymous', '/api/rpc/getItemsAnonymous'];

const validateCaptcha = async (gRecaptchaToken: string, res: IRequestMiddlewareResponse, next: MiddlewareNext) => {
  const gRecaptchaSecret = process.env.NEXT_RECAPTCHA_SECRET;
  if (!gRecaptchaToken || !gRecaptchaSecret) {
    res.status(500).json({
      status: 'failure',
      message: 'Error validating captcha'
    });
    return;
  }

  const reCaptchaResponse = await fetch('https://www.google.com/recaptcha/api/siteverify', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded'
    },
    body: `secret=${gRecaptchaSecret}&response=${gRecaptchaToken}`
  });

  const captchaResultJson = await reCaptchaResponse.json();
  if (captchaResultJson?.score <= 0.5) {
    res.status(200).json({
      status: 'failure',
      message: 'Google ReCaptcha Failure'
    });
  } else {
    await next();
  }
};

const RecaptchaMidleware: RequestMiddleware = async (
  request: IRequestMiddlewareRequest,
  response: IRequestMiddlewareResponse,
  next: MiddlewareNext
) => {
  if (protectedUrls.some((url) => url === request.url)) {
    await validateCaptcha(request.body.params.gRecaptchaToken, response, next);
  } else {
    await next();
  }
};

export default RecaptchaMidleware;
