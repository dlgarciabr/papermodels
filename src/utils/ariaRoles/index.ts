import { WIDGET } from './widget';
import { LANDMARK } from './landmark';
import { STRUCTURE } from './structure';

/**
 * Extracted from [a11y-types](https://github.com/natemoo-re/a11y-types),
 * all credits to [Nate Moore](https://github.com/natemoo-re)
 */
export const ARIA_ROLE = {
  /**
   * Widget roles act as standalone user interface widgets or as part of larger, composite widgets.
   */
  WIDGET,

  /**
   * Landmark Roles indicate regions of the page intended as navigational landmarks.
   */
  LANDMARK,
  /**
   * Structure roles describe structures that organize content in a page.
   *
   * Document structures are not usually interactive.
   */
  STRUCTURE
};
