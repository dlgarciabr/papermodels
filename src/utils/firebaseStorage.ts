import { FirebaseApp, FirebaseOptions, initializeApp, getApps } from 'firebase/app';
import { getStorage as getFirebaseStorage, listAll, ref, getDownloadURL } from 'firebase/storage';

const FILES_PATH = 'artifacts';

const getFirebaseApp = (): FirebaseApp => {
  const firebaseConfig: FirebaseOptions = {
    apiKey: 'AIzaSyDfH-Wfq2vzCT8ULhzWYbtv-6E3v3_Vw-s',
    authDomain: 'papermodels-5a9e1.firebaseapp.com',
    projectId: 'papermodels-5a9e1',
    storageBucket: 'papermodels-5a9e1.appspot.com',
    messagingSenderId: '427465803853',
    appId: '1:427465803853:web:fa782e8eb6f474f14df404',
    measurementId: 'G-B8W9NCLDY8'
  };
  if (!getApps().length) {
    // Initialize Firebase
    initializeApp(firebaseConfig);
  }
  return getApps()[0] as FirebaseApp;
};

export const getStorage = () => getFirebaseStorage(getFirebaseApp());

export const listAllFiles = async () => {
  const result = await listAll(ref(getStorage(), FILES_PATH));
  return result;
};

export const getFilePath = async (fileId) => {
  const url = await getDownloadURL(ref(getStorage(), `${FILES_PATH}/${fileId}`));
  return url;
};

export const saveImage = () => {
  throw 'Not implemented';
};

export const deleteImage = () => {
  throw 'Not implemented';
};
