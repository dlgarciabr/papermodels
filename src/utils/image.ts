/* istanbul ignore file -- @preserve */
// It's not possible to make a real test mocking document.createElement and Image
import { readAndCompressImage } from 'browser-image-resizer';

const generateThumbnailCanvas = (bytes: ArrayBuffer) =>
  new Promise<HTMLCanvasElement>((resolve) => {
    const downloadUrl = window.URL.createObjectURL(new Blob([bytes]));
    const img = document.createElement('img');
    img.onload = () => {
      const canvas = document.createElement('canvas');
      canvas.width = 100;
      canvas.height = 100;
      const ctx = canvas.getContext('2d');
      ctx?.drawImage(img, 0, 0, 100, 100);
      resolve(canvas);
    };
    img.src = downloadUrl;
  });

export const generateThumbnailBlobUrl = async (bytes: ArrayBuffer) => {
  const canvas = await generateThumbnailCanvas(bytes);
  return canvas.toDataURL();
};

export const generateThumbnailArrayBuffer = (bytes: ArrayBuffer) =>
  new Promise<ArrayBuffer>(async (resolve, reject) => {
    const canvas = await generateThumbnailCanvas(bytes);
    canvas.toBlob((blob) => {
      if (blob) {
        resolve(blob.arrayBuffer());
      } else {
        reject();
      }
    });
  });

const config = {
  quality: 0.9,
  width: 100,
  height: 100
};

export const compressImage = async (bytes: ArrayBuffer) => {
  const file = new File([bytes], 'temp');
  let resizedImage = await readAndCompressImage(file, config);
  return resizedImage;
};
