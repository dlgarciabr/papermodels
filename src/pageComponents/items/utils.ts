import { FileType } from 'db';
import { saveFile } from 'src/utils/fileStorage';
import { UploadItemFile } from './types';

export const uploadFiles = async (files: UploadItemFile[]) => {
  const promises: Promise<void>[] = [];
  files.forEach((file) => {
    file.artifactType = FileType.scheme; // TODO receive as parameter
    const index = ++file.item.files.length;
    const name = file.item.name.replaceAll(' ', '_').toLowerCase();
    const extension = file.name.split('.')[1];
    const storageName = `${file.item.id}/${name}_${file.artifactType}_${index}.${extension}`;
    file.storageName = storageName;
    promises.push(saveFile(file));
  });
  await Promise.all(promises);
};

export const saveItemFiles = async (files: UploadItemFile[], mutation: any) => {
  const promises: Promise<void>[] = [];
  files.forEach((file) => {
    promises.push(
      mutation({
        id: file.storageName,
        artifactType: file.artifactType,
        itemId: file.item.id
      })
    );
  });
  await Promise.all(promises);
};
