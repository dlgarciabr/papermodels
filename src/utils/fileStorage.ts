/* istanbul ignore file -- @preserve */
// TODO test coverage to be implemented (https://github.com/dlgarciabr/papermodels/issues/27)

import { UploadItemFile } from 'src/items/types';
import storageProvider from './storageProviders/cloudinary';

export const ARTIFACTS_PATH = process.env.NEXT_PUBLIC_STORAGE_ARTIFACTS_PATH || 'papermodel';

export const listAllFiles = async () => {
  throw new Error('not implemented');
};

// export const listAllFiles = async () => await storageProvider.listAllFiles();

/**
 * Retrieves a full downloadable artifact url from a pre-defined storage provider
 *
 * @param path initial stored artifact path
 * @returns full storage artifact url
 */
export const getFileUrl = (path: string) => storageProvider.getFileUrl(path);

export const getThumbnailUrl = (path: string) => storageProvider.getThumbnailUrl(path);

export const saveFile = (file: UploadItemFile) => storageProvider.saveFile(file, `${ARTIFACTS_PATH}/${file.item.id}`);

export const deleteFile = async (path: string) => storageProvider.deleteFile(path);
