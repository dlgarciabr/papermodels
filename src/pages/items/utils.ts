import { FileType, Item, ItemFile } from 'db';
import { deleteFile, getFilePath, saveFile } from 'src/utils/fileStorage';
import { compressImage, generateThumbnailArrayBuffer } from 'src/utils/image';
import { removeDiacritics } from 'src/utils/string';
import { UploadItemFile } from '../../items/types';

// TODO remover
export const testCallServerlessFunc = () => {
  console.log('##################testCallServerlessFunc############################');
};

export const processFiles = async (files: UploadItemFile[]) => {
  const processedFiles: UploadItemFile[] = [];
  for await (const file of files) {
    const index = ++file.item.files.length;
    const name = removeDiacritics(file.item.name).replaceAll(' ', '_').toLowerCase();
    const extension = file.name.split('.')[1];
    const storagePath = `${file.item.id}/${name}_${file.artifactType}_${index}`;
    const imageBytes = await file.arrayBuffer();

    /* istanbul ignore if @preserve */
    if (file.artifactType === FileType.preview) {
      const thumbFileName = `${storagePath}_thumb.${extension}`;
      const thumbnailBytes = await generateThumbnailArrayBuffer(imageBytes);
      const compressedThumbnail = await compressImage(thumbnailBytes);
      const thumbnailCompressedBytes = await compressedThumbnail.arrayBuffer();
      const thumbnailFile = new File([thumbnailCompressedBytes], thumbFileName) as UploadItemFile;
      thumbnailFile.artifactType = FileType.thumbnail;
      thumbnailFile.storagePath = thumbFileName;
      thumbnailFile.item = file.item;
      processedFiles.push(thumbnailFile);
    }

    const fileName = `${storagePath}.${extension}`;
    const processedFile = new File([imageBytes], fileName) as UploadItemFile;
    processedFile.artifactType = file.artifactType;
    processedFile.storagePath = fileName;
    processedFile.item = file.item;
    processedFiles.push(processedFile);
  }
  return processedFiles;
};

export const uploadFiles = (files: UploadItemFile[]) =>
  Promise.all(
    files.map(async (file) => {
      await saveFile(file);
    })
  );

export const saveItemFiles = async (files: UploadItemFile[], createFileMutation: any) => {
  for await (const file of files) {
    const index = ++file.item.files.length;
    await createFileMutation({
      storagePath: file.storagePath,
      artifactType: file.artifactType,
      itemId: file.item.id,
      index
    });
  }
};

export const sortFilesIndexes = async (
  item: Item,
  files: ItemFile[],
  updateItemFileMutation: any
): Promise<ItemFile[]> => {
  if (files.length === 0 || (files.length === 1 && files[0]?.index === 0)) {
    return files;
  }
  const lockFileStorageName = `${item.id}/.lock`;

  await saveFile(new File([], lockFileStorageName));

  const promises: Promise<void>[] = [];
  const updatedFiles: ItemFile[] = [];

  files.forEach((file, index) => {
    promises.push(
      new Promise(async (resolve) => {
        const backupFileName = `${file.storagePath}_bak`;
        const url = await getFilePath(file.storagePath);

        const response = await fetch(url, { method: 'GET' });
        const blob = await response.blob();

        await saveFile(new File([blob], backupFileName));
        await deleteFile(file.storagePath);

        const name = removeDiacritics(item.name).replaceAll(' ', '_').toLowerCase();
        const extension = file.storagePath.split('.')[1];
        const newStoragePath = `${item.id}/${name}_${file.artifactType}_${index + 1}.${extension}`;

        await saveFile(new File([blob], newStoragePath));
        await deleteFile(backupFileName);

        const updatedFile = await updateItemFileMutation({
          ...file,
          storagePath: newStoragePath,
          index
        });

        updatedFiles.push(updatedFile);
        resolve();
      })
    );
  });

  await Promise.all(promises);

  await deleteFile(lockFileStorageName);

  return updatedFiles;
};
