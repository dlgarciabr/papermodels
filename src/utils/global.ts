import { FileType, Item, ItemFile } from '.prisma/client';
import { getFilePath } from './fileStorage';

export const downloadFile = async (file: ItemFile & { url: string; item: Item }) => {
  const url = await getFilePath(file.storagePath);
  const response = await fetch(url, { method: 'GET' });
  const blob = await response.blob();

  // Create blob link to download
  // const downloadUrl = window.URL.createObjectURL(new Blob([blob]));
  const downloadUrl = window.URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = downloadUrl;
  link.setAttribute('download', file.storagePath);

  // Append to html link element page
  document.body.appendChild(link);

  // Start download
  link.click();

  // Clean up and remove the link
  (link.parentNode as any).removeChild(link);
};

export const getSimpleRandomKey = () => Math.random().toString(36).substring(2, 15);

export const getFileTypeByText = (type: string): FileType => {
  switch (type) {
    case FileType.thumbnail:
      return FileType.thumbnail;
    case FileType.instruction:
      return FileType.instruction;
    case FileType.scheme:
      return FileType.scheme;
    case FileType.preview:
    default:
      return FileType.preview;
  }
};
