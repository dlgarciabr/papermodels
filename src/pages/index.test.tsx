import { expect, vi } from "vitest";

import { render, screen } from "test/utils";
import Home from "./index";

test("renders blitz documentation link", () => {
  // arrange
  const userEmail = "user@email.com";

  vi.mock("src/users/hooks/useCurrentUser", () => ({
    useCurrentUser: () => ({
      id: 1,
      name: "User",
      email: "user@email.com",
      role: "user",
    }),
  }));

  // act
  render(<Home />);

  // assert
  expect(screen.getByText(userEmail)).toBeInTheDocument();

  const linkElement = screen.getByText(/Documentation/i);
  expect(linkElement).toBeInTheDocument();
});

test.todo("User click on logout buton");
