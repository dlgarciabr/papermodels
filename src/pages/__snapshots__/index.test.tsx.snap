// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Index page tests > Index page is rendered 1`] = `
<body
  style="overflow-y: scroll;"
>
  <div>
    <div
      class="Toastify"
    />
    <div
      class="MuiGrid-root MuiGrid-container loading css-1nwvk3e-MuiGrid-root"
      style="height: 0px;"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      />
    </div>
    <main
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item css-1h16bbz-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <img
              alt="papermodel"
              height="160px"
              layout="fixed"
              width="256px"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item css-1h16bbz-MuiGrid-root"
          style="margin-top: 46px;"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-3 MuiGrid-grid-xs-12 MuiGrid-grid-sm-10 MuiGrid-grid-md-8 MuiGrid-grid-lg-8 css-1k25aso-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 height50px css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
                style="min-height: 0px;"
              >
                <div
                  class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
                >
                  <div
                    class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardInfo MuiAlert-standard css-1fuz9qr-MuiPaper-root-MuiAlert-root"
                      role="alert"
                    >
                      <div
                        class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                          data-testid="InfoOutlinedIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
                          />
                        </svg>
                      </div>
                      <div
                        class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
                      >
                        No results were found, showing some nice suggestions!
                      </div>
                      <div
                        class="MuiAlert-action css-ki1hdl-MuiAlert-action"
                      >
                        <button
                          aria-label="Close"
                          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall css-1e0d89p-MuiButtonBase-root-MuiIconButton-root"
                          tabindex="0"
                          title="Close"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                            data-testid="CloseIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                            />
                          </svg>
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-11 css-ykovpn-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root search-input css-1z10yd4-MuiFormControl-root-MuiTextField-root"
                hidden=""
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  for=":r0:"
                  id=":r0:-label"
                >
                  Search for a model
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r0:"
                    name="searchModel"
                    type="text"
                    value=""
                  />
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
                  >
                    <button
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                      tabindex="0"
                      title="Clean"
                      type="button"
                    >
                      <svg
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M0 0h24v24H0z"
                          fill="none"
                        />
                        <path
                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-14lo706"
                    >
                      <span>
                        Search for a model
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-u78q51-MuiGrid-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge search-button css-124vkh3-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
                id=":r1:"
                tabindex="0"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="22"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="22"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title>
                    Search for a model
                  </title>
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-3 css-b2xmsj-MuiGrid-root"
        />
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-3 hidden css-b2xmsj-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 css-s2k0j8-MuiGrid-root"
          >
            <nav
              aria-label="pagination navigation"
              class="MuiPagination-root MuiPagination-text css-1oj2twp-MuiPagination-root"
            >
              <ul
                class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
              >
                <li>
                  <button
                    aria-label="Go to previous page"
                    class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                      data-testid="NavigateBeforeIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                      />
                    </svg>
                  </button>
                </li>
                <li>
                  <button
                    aria-label="Go to next page"
                    class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                      data-testid="NavigateNextIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                      />
                    </svg>
                  </button>
                </li>
              </ul>
            </nav>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 css-s2k0j8-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Total pages: 
              0
            </p>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 css-s2k0j8-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 justify-content-center height50px css-1idn90j-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              or see our categories
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 css-s2k0j8-MuiGrid-root"
          >
            <div
              class="swiper swiper-initialized swiper-horizontal"
            >
              <div
                class="swiper-wrapper"
                style="transition-duration: 0ms; transform: translate3d(NaNpx, 0px, 0px);"
              >
                <div
                  class="swiper-slide"
                >
                  <span
                    class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                    role="progressbar"
                    style="width: 40px; height: 40px;"
                  >
                    <svg
                      class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                      viewBox="22 22 44 44"
                    >
                      <circle
                        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                        cx="44"
                        cy="44"
                        fill="none"
                        r="20.2"
                        stroke-width="3.6"
                      />
                    </svg>
                  </span>
                </div>
                <div
                  class="swiper-slide"
                >
                  <span
                    class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                    role="progressbar"
                    style="width: 40px; height: 40px;"
                  >
                    <svg
                      class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                      viewBox="22 22 44 44"
                    >
                      <circle
                        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                        cx="44"
                        cy="44"
                        fill="none"
                        r="20.2"
                        stroke-width="3.6"
                      />
                    </svg>
                  </span>
                </div>
                <div
                  class="swiper-slide"
                >
                  <span
                    class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                    role="progressbar"
                    style="width: 40px; height: 40px;"
                  >
                    <svg
                      class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                      viewBox="22 22 44 44"
                    >
                      <circle
                        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                        cx="44"
                        cy="44"
                        fill="none"
                        r="20.2"
                        stroke-width="3.6"
                      />
                    </svg>
                  </span>
                </div>
              </div>
              <div
                class="swiper-button-prev"
              />
              <div
                class="swiper-button-next"
              />
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</body>
`;
